SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp

```diff
@@ -0,0 +1,170 @@
+	{
+		return;
+	}
+
+	TOptional<Trace_SpanId> SpanId = SpatialGDK::SpatialEventTracer::UserSpanIdToSpanId(UserSpanId);
+	if (!SpanId.IsSet())
+	{
+		return;
+	}
+
+	EventTracer->TraceEvent(SpatialTraceEvent, SpanId.GetValue());
+}
+
+void USpatialEventTracerUserInterface::TraceRPC(UObject* WorldContextObject, FEventTracerRPCDelegate Delegate,
+												const FUserSpanId& UserSpanId)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	if (!UserSpanId.IsValid())
+	{
+		return;
+	}
+
+	TOptional<Trace_SpanId> SpanId = SpatialGDK::SpatialEventTracer::UserSpanIdToSpanId(UserSpanId);
+	if (!SpanId.IsSet())
+	{
+		return;
+	}
+
+	EventTracer->SpanIdStack.Stack(SpanId.GetValue());
+	Delegate.Execute();
+	EventTracer->SpanIdStack.Pop();
+}
+
+bool USpatialEventTracerUserInterface::GetActiveSpanId(UObject* WorldContextObject, FUserSpanId& OutUserSpanId)
+{
+	const SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return false;
+	}
+
+	if (!EventTracer->SpanIdStack.HasSpanId())
+	{
+		return false;
+	}
+
+	TOptional<Trace_SpanId> SpanId = EventTracer->SpanIdStack.GetSpanId();
+	if (!SpanId.IsSet())
+	{
+		return false;
+	}
+
+	OutUserSpanId = SpatialGDK::SpatialEventTracer::SpanIdToUserSpanId(SpanId.GetValue());
+	return true;
+}
+
+void USpatialEventTracerUserInterface::TraceProperty(UObject* WorldContextObject, UObject* Object, const FUserSpanId& UserSpanId)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	USpatialNetDriver* NetDriver = GetSpatialNetDriver(WorldContextObject);
+	if (NetDriver == nullptr)
+	{
+		return;
+	}
+
+	TOptional<Trace_SpanId> SpanId = SpatialGDK::SpatialEventTracer::UserSpanIdToSpanId(UserSpanId);
+	if (!SpanId.IsSet())
+	{
+		return;
+	}
+
+	EventTracer->AddLatentPropertyUpdateSpanId(Object, SpanId.GetValue());
+}
+
+SpatialGDK::SpatialEventTracer* USpatialEventTracerUserInterface::GetEventTracer(UObject* WorldContextObject)
+{
+	const USpatialNetDriver* NetDriver = GetSpatialNetDriver(WorldContextObject);
+	if (NetDriver == nullptr || NetDriver->Connection == nullptr)
+	{
+		UE_LOG(LogSpatialEventTracerUserInterface, Error,
+			   TEXT("USpatialEventTracerUserInterface::GetEventTracer - NetDriver or Connection is null"));
+		return nullptr;
+	}
+
+	return NetDriver->Connection->GetEventTracer();
+}
+
+USpatialNetDriver* USpatialEventTracerUserInterface::GetSpatialNetDriver(UObject* WorldContextObject)
+{
+	const UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::ReturnNull);
+	if (World == nullptr)
+	{
+		UE_LOG(LogSpatialEventTracerUserInterface, Error, TEXT("USpatialEventTracerUserInterface::GetSpatialNetDriver - World is null"));
+		World = GWorld;
```

Comment:
Why is this required?

Reply:
Tbh, maybe it isn't... I copied it from a similar test I believe

Reply:
Oh the latency tracer... haha

Reply:
what is GWorld? Is it guaranteed to have a connection/tracer!?

Reply:
I have no idea tbh. If it does not have a connection or tracer we just return nullptrs.